### DISCUSSION TITLE
visit API no longer offers parents and visitParents doesn't have parents?

### DISCUSSION THREAD

**Discussion by kentcdodds** [syntax-tree/unist#118] [Q&A]
Here's my current code that worked with `unist-util-visit` v4.1.2: https://github.com/kentcdodds/kentcdodds.com/blob/95663825007397455c634ae1527756377248302a/app/utils/simplecast.server.ts#L296

Now I'm trying to upgrade and I don't see a migration guide, but apparently `index` and `parents` are no longer supplied. So I'm trying `unist-util-visit-parents`:

```tsx
        visitParents(tree, 'heading', (heading: M.Heading, parents) => {
          const [parent] = parents
          if (!parent) {
            console.error(heading, `${errorKey} heading without a parent`)
            return
          }
          // ...
```

And in this case `parents` is `Array<never>`.

I'm not sure what I'm doing wrong.

**Comment by remcohaszing** [comment]
The unkfied ecosystem is being updated with (mostly small) breaking changes. The unist utilities have already been updated, but remark not yet. You're best off updating `unist-util-*` after remark has been updated.

**Reply by kentcdodds** [reply]
Oh, ok. I probably don't need to explain why this is frustrating and unexpected. It would be nice if these sorts of inter-dependent breaking changes were more coordinated somehow.

**Reply by wooorm** [reply]
Yes, it’s annoying!

This “actually” still works though, but doesn’t work in TS; it’s a breaking change to the types.
We have so many projects that deal with the ASTs, I need to update every package, which takes a lot of time. I’m almost done though, only remark is left. Probably end of the month!

It’s one of the downsides of TS IMO. It has a lot of benefits. But a change like this, where basically everything still works but the types don’t, is heavy.

Let me know if you have ideas on how to improve this!

**Reply by kentcdodds** [reply]
I'm sure you've already considered a monorepo and have reasons for not adopting it, but I think this is a big part of why they exist

**Reply by wooorm** [reply]
We could use more monorepos! Some of the reasoning behind why it’s this way is more historical, it could change.

But, we do have already have several monorepos, and when I upgrade them, I notice I’m spending *a lot* of time on them.
More monorepos probably will shave time off, but I don’t think it’s that significant. You’d still have say 4 test suites and 4 readmes and 4 times the regular code.
On the other hand, for a simple project in a non-monorepo, I can get through it very fast.
